File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        if not root or not p or not q:
            return None
        res = root
        root_stack = [root]
        while root_stack:
            curr_node = root_stack.pop()
            res = curr_node
            #if curr_node.val == p.val or curr_node.val == q.val or not (curr_node.val < p.val and curr_node.val < q.val) or not (curr_node.val > p.val and curr_node.val > q.val):
            #    break
            if curr_node.val < p.val and curr_node.val < q.val:
                root_stack.append(curr_node.right)
            elif curr_node.val > p.val and curr_node.val > q.val:
                root_stack.append(curr_node.left)
            else:
                break
        return res

File: Data Structures & Algorithms/house-robber-iii/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    pair<int, int> dfs(TreeNode* node){
        if (node == nullptr){
            return {0, 0};
        }
        pair<int, int> left = dfs(node->left);
        pair<int, int> right = dfs(node->right);
        int firstLevel = left.second + right.second + node->val;
        int secondLevel = max(left.second, left.first) + max(right.second, right.first);
        return {firstLevel, secondLevel};
    }
public:
    int rob(TreeNode* root) {
        pair<int, int> res = dfs(root);
        return max(res.first, res.second);
    }
};